src/Cake.Core/CakeTask.cs

```diff
@@ -35,165 +30,114 @@ public abstract class CakeTask : ICakeTaskInfo
-
-            _dependencies = new List<CakeTaskDependency>();
-            _reverseDependencies = new List<CakeTaskDependency>();
-            _criterias = new List<Func<ICakeContext, bool>>();
-
-            Name = name;
-        }
+        /// <value>The task's actions.</value>
+        public List<Func<ICakeContext, Task>> Actions { get; }
 
         /// <summary>
-        /// Adds a dependency to the task.
+        /// Gets the task's actions that are run at execution time to additionally populate <see cref="Actions"/>.
         /// </summary>
-        /// <param name="name">The name of the dependency.</param>
-        /// <param name="required">Whether or not the dependency is required.</param>
-        public void AddDependency(string name, bool required = true)
-        {
-            if (_dependencies.Any(x => x.Name == name))
-            {
-                const string format = "The task '{0}' already have a dependency on '{1}'.";
-                var message = string.Format(CultureInfo.InvariantCulture, format, Name, name);
-                throw new CakeException(message);
-            }
-            _dependencies.Add(new CakeTaskDependency(name, required));
-        }
+        /// <value>The task's delayed actions actions.</value>
+        public Queue<Action> DelayedActions { get; }
 
         /// <summary>
-        /// Makes this task a dependency of some other task.
+        /// Gets or sets a value indicating whether gets the task's state if it will defer exceptions until the end of the task.
         /// </summary>
-        /// <param name="name">The name of the task that this task want to be a dependency of.</param>
-        /// <param name="required">Whether or not the dependency is required.</param>
-        public void AddReverseDependency(string name, bool required = true)
-        {
-            if (_reverseDependencies.Any(x => x.Name == name))
-            {
-                const string format = "The task '{0}' already is a dependee of '{1}'.";
-                var message = string.Format(CultureInfo.InvariantCulture, format, Name, name);
-                throw new CakeException(message);
-            }
-            _reverseDependencies.Add(new CakeTaskDependency(name, required));
-        }
+        /// <value>The task's defer exceptions state.</value>
+        public bool DeferExceptions { get; set; }
 
-        /// <summary>
-        /// Adds a criteria to the task that is invoked when the task is invoked.
-        /// </summary>
-        /// <param name="criteria">The criteria.</param>
-        public void AddCriteria(Func<ICakeContext, bool> criteria)
-        {
-            if (criteria == null)
-            {
-                throw new ArgumentNullException(nameof(criteria));
-            }
-            _criterias.Add(criteria);
-        }
+        IReadOnlyList<CakeTaskDependency> ICakeTaskInfo.Dependencies => Dependencies;
+        IReadOnlyList<CakeTaskDependency> ICakeTaskInfo.Dependees => Dependees;
```

Comment:
Hmmm, so the interface exposes `IReadOnlyList<T>` but the class exposes `List<T>`. What was the reason for this? Should we instead change the type in the interface? (While we are breaking things 😄 )

Reply:
@mholo65 The interface exposes a readonly view of the task and shouldn't be changed for backwards compatibility.